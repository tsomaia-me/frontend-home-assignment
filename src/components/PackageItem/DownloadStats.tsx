import React, { useMemo } from 'react'
import 'chart.js/auto'
import { Chart } from 'react-chartjs-2'
import { useEnvironment } from '../../hooks'
import { DateTime } from 'luxon'

export interface DownloadStatsProps {
  downloadStats: Array<number>
}

function DownloadStats({ downloadStats }: DownloadStatsProps) {
  const { numberOfDownloadStatsDays } = useEnvironment()
  const labels = useMemo(
    () => Array(numberOfDownloadStatsDays)
      .fill(null)
      .map((_, index) => DateTime.now().minus({ day: numberOfDownloadStatsDays - index + 1 }).toFormat('ccc')),
    [numberOfDownloadStatsDays]
  )
  const totalCount = useMemo(() => downloadStats.reduce((a, b) => a + b, 0), [downloadStats])

  return (
    <div>
      <h3>{downloadStats.length === 0 ? '--' : totalCount} downloads in the last week</h3>
      <Chart
        type="line"
        data={{
          labels,
          datasets: [
            {
              label: '',
              data: downloadStats,
              borderColor: '#3f51b5',
            }
          ]
        }}
      />
    </div>
  )
}

export default DownloadStats
