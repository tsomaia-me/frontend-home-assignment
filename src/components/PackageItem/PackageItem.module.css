.root {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
}

.sidebar {
    min-width: 300px;
    width: 40%;
}

.details {
    margin: 0;
    padding: 0;
    list-style-type: none;
    font-size: 18px;
}

.details > li {
    display: flex;
    flex-direction: row;
    align-items: center;
    padding: 16px 0;
    border-bottom: 1px solid #e0e0e0;
}

.details > li > :last-child:not(:first-child) {
    margin-left: 8px;
}

.details > li > a, .details > li > a:visited, .details > li > a:active {
    color: #2196f3;
    font-weight: bold;
    text-decoration: none;
}

.details > li > a:hover {
    text-decoration: underline;
}
