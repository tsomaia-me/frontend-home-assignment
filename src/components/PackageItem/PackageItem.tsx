import React from 'react'

import styles from './PackageItem.module.css'
import { PackageDetails } from '../../models'
import DownloadStats from './DownloadStats'

export interface PackageProps {
  item: PackageDetails
}

function PackageItem({ item }: PackageProps) {
  return (
    <article className={styles.root}>
      <div>
        <h2>{item.package.name}</h2>
        <div>
          {item.package.description || 'No description available for this package.'}
        </div>
      </div>

      <aside className={styles.sidebar}>
        <ul className={styles.details}>
          {item.package.homepage && (
            <li>
              <span className="material-icons">link</span>
              <a href={item.package.homepage?.toString()} target="_blank" rel="noreferrer">
                {item.package.homepage?.toString()}
              </a>
            </li>
          )}
          {item.package.repository?.url && (
            <li>
              <span className="material-icons">code</span>
              <a href={item.package.repository.url} target="_blank" rel="noreferrer">
                {item.package.repository.url}
              </a>
            </li>
          )}
          {item.package.license && (
            <li>
              <span className="material-icons">balance</span>
              <span>{item.package.license}</span>
            </li>
          )}
        </ul>
        <DownloadStats downloadStats={item.downloadStats}/>
      </aside>
    </article>
  )
}

export default PackageItem
