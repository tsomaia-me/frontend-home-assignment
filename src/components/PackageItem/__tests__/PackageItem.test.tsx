import React from 'react'
import renderer from 'react-test-renderer'
import PackageItem from '../PackageItem'

test('renders PackageItem correctly', () => {
  const json = renderer.create(
    <PackageItem
      item={{
        package: { name: 'test', description: 'test description', dependencies: {} },
        downloadStats: [],
      }}
    />
  ).toJSON()
  expect(json).toMatchSnapshot()
})
