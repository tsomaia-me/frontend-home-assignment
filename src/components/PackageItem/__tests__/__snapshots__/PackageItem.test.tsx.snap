// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders PackageItem correctly 1`] = `
<article
  className="root"
>
  <div>
    <h2>
      test
    </h2>
    <div>
      test description
    </div>
  </div>
  <aside
    className="sidebar"
  >
    <ul
      className="details"
    />
    <div>
      <h3>
        --
         downloads in the last week
      </h3>
      <canvas
        height={150}
        role="img"
        width={300}
      />
    </div>
  </aside>
</article>
`;
